tc/core/compiler_api_rfc.h

```diff
@@ -0,0 +1,127 @@
+/// options then just compile another TcExecutor; because atm we fully JIT
+/// specialize on all sizes.
+/// \returns a new TcExecutor on which the run method can be called to run
+/// entryPoint
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options);
+
+/// Given a TC representation as a TC + TC function name entryPoint and a list
+/// of input tensors that match the definition in the TC function definition
+/// (in positional order), this generates the output TensorInfo resulting from
+/// running inference.
+/// The typical flow is to infer output sizes, allocate/resize them within
+/// you favorite ML framework/tensor library and then call compile and run.
+/// \returns a vector of TensorInfo which can be used for allocating and
+/// performing output shape validation.
+std::vector<TensorInfo> inferOutputTensorInfo(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*> inputs);
+
+namespace detail {
+/// Given a TC representation as a TreeRef, this function compiles a new
+/// TcExecutor for the specified Backend.
+/// For now, contiguous output sizes are inferred given input sizes.
+/// If you need another kernel for another TC or another inputs, options then
+/// just compile another TcExecutor; because atm we fully JIT specialize on all
+/// sizes.
+/// \returns a new TcExecutor on which the run method can be called
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    lang::TreeRef tcDefinition,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options);
+
+/// Given a TC representation as a TreeRef and a list of input tensors that
+/// match the definition in the TC function definition (in positional order),
+/// this generates the output TensorInfo resulting from running inference.
+/// The typical flow is to infer output sizes, allocate/resize them within
+/// you favorite ML framework/tensor library and then call compile and run.
+/// \returns a vector of TensorInfo which can be used for allocating and
+/// performing output shape validation.
+std::vector<TensorInfo> inferOutputTensorInfo(
+    lang::TreeRef tcDefinition,
+    const std::vector<const DLConstTensor*> inputs);
+
+/// Given a TC representation, this parses the TC functions into a map of
+/// TreeRef indexed by TC function names.
+/// \returns an ordered map of TC function name to parsed TC tree
+std::map<std::string, lang::TreeRef> parse(const std::string& tc);
+} // namespace detail
+} // namespace tc
```

Comment:
Nit: put `parse` declaration before `compile` declaration since the latter
requires the output of the former.